sematic/resolvers/state_machine_resolver.py

```diff
@@ -18,37 +19,52 @@ def __init__(self, detach: bool = False):
         self._futures: typing.List[AbstractFuture] = []
         self._detach = detach
 
+    @property
+    def _root_future(self) -> AbstractFuture:
+        return self._futures[0]
+
     def resolve(self, future: AbstractFuture) -> typing.Any:
-        resolved_kwargs = self._get_resolved_kwargs(future)
-        if not len(resolved_kwargs) == len(future.kwargs):
-            raise ValueError(
-                "All input arguments of your root function should be concrete."
-            )
+        try:
+            resolved_kwargs = self._get_resolved_kwargs(future)
+            if not len(resolved_kwargs) == len(future.kwargs):
+                raise ValueError(
+                    "All input arguments of your root function should be concrete."
+                )
 
-        future.resolved_kwargs = resolved_kwargs
+            future.resolved_kwargs = resolved_kwargs
 
-        self._resolution_will_start()
+            self._enqueue_future(future)
 
-        self._enqueue_future(future)
+            if self._detach:
+                return self._detach_resolution(future)
 
-        if self._detach:
-            return self._detach_resolution(future)
+            self._resolution_will_start(future)
 
-        while future.state != FutureState.RESOLVED:
-            for future_ in self._futures:
-                if future_.state == FutureState.CREATED:
-                    self._schedule_future_if_args_resolved(future_)
-                if future_.state == FutureState.RAN:
-                    self._resolve_nested_future(future_)
+            while future.state != FutureState.RESOLVED:
+                for future_ in self._futures:
+                    if future_.state == FutureState.CREATED:
+                        self._schedule_future_if_args_resolved(future_)
+                    if future_.state == FutureState.RAN:
+                        self._resolve_nested_future(future_)
 
-            self._wait_for_scheduled_run()
+                self._wait_for_scheduled_run()
 
-        self._resolution_did_succeed()
+            self._resolution_did_succeed()
 
-        if future.state != FutureState.RESOLVED:
-            raise RuntimeError("Unresolved Future after resolver call.")
+            if future.state != FutureState.RESOLVED:
+                raise RuntimeError("Unresolved Future after resolver call.")
 
-        return future.value
+            return future.value
+        except Exception as e:
+            due_to_calculator_error = False
+            if isinstance(e, CalculatorError):
+                due_to_calculator_error = True
+            self._resolution_did_fail(error=e)
+            if due_to_calculator_error and hasattr(e, "__cause__"):
```

Comment:
Why the need for `due_to_calculator_error`? Why not just
```python
if isinstance(e, CalculatorError) and hasattr(e, "__cause__"):
```

Reply:
done